lisa/runners/lisa_runner.py

```diff
@@ -20,298 +21,384 @@
-    ) -> None:
-
-        assert case_results
-        suite_metadata = case_results[0].runtime_data.metadata.suite
+        assert test_results
+        suite_metadata = test_results[0].runtime_data.metadata.suite
         test_suite: TestSuite = suite_metadata.test_class(
             suite_metadata,
         )
-        test_suite.start(environment=environment, case_results=case_results)
+        test_suite.start(environment=environment, case_results=test_results)
+
+    def _delete_environment_task(
+        self, environment: Environment, test_results: List[TestResult] = []
+    ) -> None:
+        """
+        May be called async
+        """
+        # the predefined environment shouldn't be deleted, because it
+        # serves all test cases.
+        if (
+            environment.status
+            in [
+                EnvironmentStatus.Deployed,
+                EnvironmentStatus.Connected,
+            ]
+        ) or (
+            environment.status == EnvironmentStatus.Prepared and environment.is_in_use
+        ):
+            self.platform.delete_environment(environment)
+        else:
+            environment.status = EnvironmentStatus.Deleted
+
+    def _get_same_priority_results(
+        self, test_results: List[TestResult]
+    ) -> List[TestResult]:
+        if not test_results:
+            return []
+
+        current_priority = test_results[0].runtime_data.metadata.priority
+        test_results = [
+            x
+            for x in test_results
+            if x.runtime_data.metadata.priority == current_priority
+        ]
+        while not test_results:
+            current_priority += 1
+            test_results = [
+                x
+                for x in test_results
+                if x.runtime_data.metadata.priority == current_priority
+            ]
+
+        return test_results
+
+    def _generate_task(
+        self,
+        task_method: Callable[[Environment, List[TestResult]], None],
+        environment: Environment,
+        test_results: List[TestResult],
+    ) -> Callable[[], List[TestResult]]:
+        assert not environment.is_in_use
+        environment.is_in_use = True
+        for test_result in test_results:
+            # return assigned but not run casese
+            if test_result.status == TestStatus.QUEUED:
+                test_result.set_status(TestStatus.ASSIGNED, "")
+
+        task = partial(self._run_task, task_method, environment, test_results)
+        return task
+
+    def _run_task(
+        self,
+        task_method: Callable[[Environment, List[TestResult]], None],
+        environment: Environment,
+        test_results: List[TestResult],
+    ) -> List[TestResult]:
+        assert environment.is_in_use
+        task_method(environment, test_results)
+
+        for test_result in test_results:
+            # return assigned but not run casese
+            if test_result.status == TestStatus.ASSIGNED:
+                test_result.set_status(TestStatus.QUEUED, "")
```

Comment:
Wait, then tests get back to queued? Thas makes no sense

Reply:
Think it's a typo, it should be RUNNING?

Reply:
The ASSIGNED status is used to assign a test case to an environment exclusively. It prevents one test result is assigned to two environment, it causes concurrency issue. It doesn't mean this environment is the only chance of a test case to run.
There are some reasons that an environment doesn't run or use a test result. For example, during deployment, initialization phase, only if it's failed, the test case will be failed, and the environment error attached on it. If multiple test results are assigned to an environment, but environment fails, the second and other test results will be returned to queue for next try. Some test case can throw queued exception to try next environment dynamically...

Reply:
ok may be I am missing something. So if some (not all) test_result.status is still ASSIGNED, then we set them to QUEUED but why we are marking environment.is_in_use as False?
Can't there be some test_result.status as "RUNNING"?

Reply:
And what's the intention for tests being attachable to multiple environments, in real life? Something like multiple Azure regions? Just one example would help :)

Reply:
> ok may be I am missing something. So if some (not all) test_result.status is still ASSIGNED, then we set them to QUEUED but why we are marking environment.is_in_use as False?
> Can't there be some test_result.status as "RUNNING"?

The test results attaches to an environment to run. Once it's done, not run test results will be put back to queue, and the environment will be released (is_in_used=false) for next batch of test results.

Reply:
> And what's the intention for tests being attachable to multiple environments, in real life? Something like multiple Azure regions? Just one example would help :)

If a test case doesn't specify requirement, it can be run in most environments. If some environment fails to deploy, the assigned test cases can try another one, which fits its requirement, instead of fail them directly. It gives bigger coverage. For example, current three test cases (smoke test, lsvmbus, timesync), they have no special requirement on environments.

Reply:
Right, let's continue on that. What if the user wants to run them in all environments? What are the real use cases for > 1 environment at the same time, speaking Azure?

Reply:
Run 1 case on multiple environments (like vm_size) are handled by test matrix, the test purpose is on vm_size or other dimensions. If we are validating a distro, it needs to validate it with different environments (like SRIOV on/off, different disk types). Both of them are used by current test scenarios, and supported in V2.